### PR TITLE
Implement scss variables

### LINKED ISSUE #2288: Change themed CSS generation to SCSS templates
Thanks to the introduction of Sass in nextcloud (#2050), we could really improve how the css styles are generated in the Theming Controller, because now the controller is bloated with concatenated CSS strings: [server/apps/theming/lib/Controller/ThemingController.php](https://github.com/nextcloud/server/blob/master/apps/theming/lib/Controller/ThemingController.php)

We could use the [scssphp](http://leafo.net/scssphp/) compiler to compile SASS templates with variables:
```php
echo $scss->compile('
  $color: #abc;
  div { color: lighten($color, 20%); }
');
```
This would also make theming more accessible.
What do you think @juliushaertl, @LukasReschke, @schiessle @nickvergessen ?

### PR DESCRIPTION
**Pull request by juliusknorr**
This is a basic implementation for using SCSS variables. For now I only intruduced variables for colors. We can add others for border-radius, etc. later. 

fix #2288 #3256 #3211 

@skjnldsv @nextcloud/designers Please have a look if it makes sense so far or if anything important is missing.

### Applied to files
- [x] header.scss 
- [x] apps.scss
- [x] icons.scss  
- [x] inputs.scss  
- [x] multiselect.scss  
- [x] share.scss 
- [x] styles.scss  
- [x] systemtags.scss  
- [x] tooltip.scss

### CODE REVIEW

**Comment by skjnldsv** [comment]
This is looking soooo goooood! 💟

**Comment by raimund-schluessler** [comment]
Sorry if this is the wrong place, but I haven't found anything in the docs or in the respective issues, so:
Can we also use SCSS for styling apps?

**Comment by skjnldsv** [comment]
@raimund-schluessler https://github.com/nextcloud/server/pull/3197

**Comment by raimund-schluessler** [comment]
Great, thanks, I should have looked in the PRs too.

**Comment by MorrisJobke** [comment]
> Sorry if this is the wrong place, but I haven't found anything in the docs or in the respective issues, so:
> Can we also use SCSS for styling apps?

Good point. @skjnldsv Can you add a note about this in the dev docs?

File: lib/private/Template/SCSSCacher.php [hunk 1]
```diff
@@ -114,6 +113,27 @@ private function isCached($fileNameCSS, $fileNameSCSS, ISimpleFolder $folder, $p
 	}
 
 	/**
+	 * Check if the variables file has changed
+	 * @param $fileNameCSS
```
Comment (LukasReschke) [inline lib/private/Template/SCSSCacher.php]: string? potatoes? ananas? array? ;)

Comment (skjnldsv) [inline lib/private/Template/SCSSCacher.php]: 🍍 👍

Comment (juliushaertl) [inline lib/private/Template/SCSSCacher.php]: Hmm now I'm hungry. :open_mouth:

**Comment by juliushaertl** [comment]
I've now replaced all the colors with temporary variables, so we can better track which color it was before. Next step would be to reduce those values to a basic set of variables: https://github.com/nextcloud/server/blob/e3a621d7608d726b63c99f3660c29debb9e29ba8/core/css/variables.scss#L1-L10

I'm not sure how we should handle the simplification of the massive amount of grey values:
https://github.com/nextcloud/server/blob/e3a621d7608d726b63c99f3660c29debb9e29ba8/core/css/variables.scss#L21-L43

@skjnldsv Using darken/lighten function might be ok, but when it comes to theming (especially themes with inverted colors) using rgba($color-main-text, 0.1) for the different grey variations might be better. What do you think?

**Comment by skjnldsv** [comment]
Couldn't we set the darken/lighten functions into the variable.scss file?
So someone can easily edit this?

**Comment by skjnldsv** [comment]
For the grey values, we should merge what we have into simpler ones. But this could require to look exactly where they are used since they may be erroneous already. I can see someone putting a #eee right next a #e6e6e6 without seeing what he was doing on the first place :p

File: lib/private/Template/SCSSCacher.php (hunk 1 above)
Comment (MorrisJobke) [inline lib/private/Template/SCSSCacher.php]: 🍕

File: core/css/variables.scss [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+// need to be reduced/replaced
+$color-main-old-ff0000: $color-error;
+$color-main-old-ffdddd: rgba($color-error, .15);
+
+$color-main-old-fafafa: #fafafa;
+$color-main-old-f8f8f8: #f8f8f8;
+$color-main-old-245: rgb(245,245,245);
+$color-main-old-f0f0f0: f0f0f0;
+$color-main-old-238: rgb(238, 238, 238);
```
Comment (MorrisJobke) [inline core/css/variables.scss]: Same as `#eee`

File: core/css/variables.scss [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+$image-login-background: '../img/background.jpg?v=1';
+
+// need to be reduced/replaced
+$color-main-old-ff0000: $color-error;
+$color-main-old-ffdddd: rgba($color-error, .15);
+
+$color-main-old-fafafa: #fafafa;
+$color-main-old-f8f8f8: #f8f8f8;
+$color-main-old-245: rgb(245,245,245);
+$color-main-old-f0f0f0: f0f0f0;
```
Comment (MorrisJobke) [inline core/css/variables.scss]: Missing `#`?

File: core/css/variables.scss [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+$color-main-old-238: rgb(238, 238, 238);
+$color-main-old-eeeeee: #eee;
+$color-main-old-e8e8e8: #e8e8e8;
+$color-main-old-dddddd: #ddd;
+$color-main-old-220: rgb(220, 220, 200);
```
Comment (MorrisJobke) [inline core/css/variables.scss]: It's not really grey, but slightly green `#DCDCC8` - But I can't find any reference in the above code.

File: core/css/variables.scss [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+$color-main-old-d3d3d3: #d3d3d3;
+$color-main-old-cccccc: #ccc;
+$color-main-old-187: rgb(198, 198, 198);
```
Comment (MorrisJobke) [inline core/css/variables.scss]: `#C6C6C6`

File: core/css/variables.scss [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+$color-main-old-190: rgb(190, 190, 190);
```
Comment (MorrisJobke) [inline core/css/variables.scss]: `#BEBEBE`

File: core/css/variables.scss [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+$color-main-old-bbbbbb: #bbb;
+$color-main-old-aaaaaa: #aaa;
+$color-main-old-999999: #999;
+$color-main-old-150: rgb(150, 150, 150);
```
Comment (MorrisJobke) [inline core/css/variables.scss]: `#969696`

File: core/css/variables.scss [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+$color-main-old-888888: #888;
+$color-main-old-777777: #777;
+$color-main-old-666666: #666;
+$color-main-old-555555: #555;
+$color-main-old-222222: #222;
+$color-main-old-333333: #333333;
```
Comment (MorrisJobke) [inline core/css/variables.scss]: Sorted list of grey values:

 ```
#222222
#333333
#555555
#666666
#777777
#888888
#969696
#999999
#aaaaaa
#bbbbbb
#bebebe
#c6c6c6
#cccccc
#d3d3d3
#dddddd
#e8e8e8
#eeeeee
#eeeeee
#f0f0f0
#f5f5f5
#f8f8f8
#fafafa
```

File: core/css/variables.scss (hunk 6 above)
Comment (skjnldsv) [inline core/css/variables.scss]: 🐑

File: core/css/variables.scss (hunk 8 above)
Comment (MorrisJobke) [inline core/css/variables.scss]: Maybe we in a first step remove similar colours, like those:

bildschirmfoto 2017-02-20 um 12 27 06

**Comment by juliushaertl** [comment]
@MorrisJobke @LukasReschke @skjnldsv @nextcloud/designers Ready for review.

We can squash those commits, but I think it might be easier to trace what has changed exactly for reviewing.

**Review by MorrisJobke — APPROVED** [review]
Tested and works nicely 👍

**Review by LukasReschke — APPROVED** [review]
Works nicely :)

**Comment by MorrisJobke** [comment]
🎉 Thanks a lot @juliushaertl

**Pull request merged**